stdlib/LinearAlgebra/src/LinearAlgebra.jl

```diff
@@ -516,32 +516,55 @@ const ⋅ = dot
 const × = cross
 export ⋅, ×
 
+# Separate the char corresponding to the wrapper from that corresponding to the uplo
+# In most cases, the former may be constant-propagated, while the latter usually can't be.
+# This improves type-inference in wrap for Symmetric/Hermitian matrices
+# A WrapperChar is equivalent to `isuppertri ? uppercase(wrapperchar) : lowercase(wrapperchar)`
+struct WrapperChar <: AbstractChar
+    wrapperchar :: Char
+    isuppertri :: Bool
+end
+function Base.Char(w::WrapperChar)
+    T = w.wrapperchar
+    if T ∉ ('S', 'H')
+        T
+    else
+        _isuppertri(w) ? uppercase(T) : lowercase(T)
+    end
+end
+Base.codepoint(w::WrapperChar) = codepoint(Char(w))
+WrapperChar(n::UInt32) = WrapperChar(Char(n), true)
+# We extract the wrapperchar so that the result may be constant-propagated
+# This doesn't return a value of the same type on purpose
+Base.uppercase(w::WrapperChar) = uppercase(w.wrapperchar)
+Base.lowercase(w::WrapperChar) = lowercase(w.wrapperchar)
+_isuppertri(w::WrapperChar) = w.isuppertri
+_isuppertri(x::AbstractChar) = isuppercase(x) # compatibility with earlier Char-based implementation
+_getuplo(x) = _isuppertri(x) ? (:U) : (:L)
+
 wrapper_char(::AbstractArray) = 'N'
```

Comment:
I'm a bit unsure if all of these should be `WrapperChar`s. On one hand, using a `Char` here simplifies the comparison to another char, and the result may usually be constant-propagated. On the other hand, this may make `tA` and `tB` have different types, which might lead to certain results be inferred as a `Union` instead of as a concrete type. Perhaps this is ok, as union-splitting should be able to handle this.

Reply:
I can't judge this. Can you observe something special when you multiply a `Symmetric` with a `Matrix`?

Reply:
This doesn't seem to matter in some tests that I ran, so I'll leave this as it is now. This way, we make the minimal changes that are necessary to get constant-propagation working for `Symmetric`/`Hermitian`, while the other types remain untouched.